Detection (#16)
Imageomics/hpc-inference: docs/animal-detection-guide.md

# Animal Detection Guide

This guide covers how to perform animal detection on large image datasets using MegaDetector[^1] [^2]. MegaDetector is specifically trained for wildlife camera trap images.

[^1]: Beery, S., Morris, D., & Yang, S. MegaDetector. GitHub repository: https://github.com/agentmorris/MegaDetector

[^2]: Microsoft AI for Earth. CameraTraps: Tools for training and running detectors and classifiers for wildlife camera trap images. GitHub repository: https://github.com/microsoft/CameraTraps

## Overview

The animal detection module provides high-performance batch processing of images for wildlife detection using MegaDetector. It's designed for HPC environments with GPU acceleration and distributed processing capabilities.

### Key Features

> - **MegaDetector Support**: Optimized for Microsoft's MegaDetector models (specifically designed for wildlife camera trap images)

Reviewer: I'd just add it to the citation further up (since you cite Megadetector with both already). Also, change:
```suggestion
- **MegaDetector Support**: Optimized for [Pytorch-Wildlife-model-weights](https://zenodo.org/records/15398270), specifically, the Megadetector models designed for wildlife camera trap images.
```

}
```

### Example Output

```python
import pandas as pd
import json

# Load results
df = pd.read_parquet("~/animal_detection/detection/detections/rank_0/animal_detection_results_rank_0_0.parquet")

# View summary
print(f"Processed {len(df)} images")
print(f"Found animals in {sum(df['num_detections'] > 0)} images")
print(f"Average detections per image: {df['num_detections'].mean():.2f}")
```
```
Processed 1120 images
Found animals in 1116 images
Average detections per image: 1.10
```
```python
# Parse detections for first image with animals
first_detection = df[df['num_detections'] > 0].iloc[0]
detections = json.loads(first_detection['detections'])
print(f"Image {first_detection['uuid']} has {len(detections)} detections:")
for detection in detections:
    print(f"  {detection['class_name']}: {detection['confidence']:.3f}")
```
```
[{'bbox': [188.17922973632812,
   656.734619140625,
   443.3848876953125,
   794.3240966796875],
  'bbox_normalized': [0.14701502323150634,
   0.5130739212036133,
   0.3463944435119629,
   0.6205657005310059],
  'confidence': 0.8981033563613892,
  'class_id': 0,
  'class_name': 'animal'}]
```

> Megadetector detected one animal with high confident 0.8981. Let's use the utility function to plot the detection box on top of the original image to validate the detection result:

Reviewer: ```suggestion
Megadetector detected one animal with high confidence (0.8981). Let's use the utility function to plot the detection box on top of the original image to validate the detection result:
```

```python
from hpc_inference.utils.visualization import plot_detections_matplotlib

image_uuid = first_detection['uuid']
image_path = f"camera_trap_data/data/test/ENA24/{image_uuid}"

image = Image.open(image_path)
detections = json.loads(first_detection['detections'])

plot_detections_matplotlib(
    image, detections,
    letterbox_shape=(1280, 1280),
    title=f"Detections for {image_uuid}"
)
```

> ![Original vs Processed Beetle Images](imgs/detection_box.png)

Reviewer: ```suggestion
![Image 00170382-1e3d-4d46-9243-ac78136922ba.png with bounding box overlay for detection, original image from](imgs/detection_box.png)
```
Add in source for original image. May want to put in a caption or something.
